### PR TITLE
Skipping tests if AZUL_DSS_ENDPOINT is not defined (#2496) 

### PR DESCRIPTION
**Pull request by amarjandu**
Author (check before every review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `DCP/1` label                                   <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes          <sub>or this PR does not need porting</sub> 
- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain            <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox`                             <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox`                                   <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain                                       <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations                        <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Reindexed `dev`                                       <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

**Review by achave11 — APPROVED** [review]
Looks good! The reduction in coverage is a bit unsettling, but since a part of the test are not running do to a DSS endpoint not being set, it makes sense.

**Comment by amarjandu** [comment]
Confirmed IT passes on 8a1c028a8bc157305c4261199213d506fdffa5b3

**Comment by hannes-ucsc** [comment]
1. Push commit #1. Sandbox will fail in test job.
2. Push commits #1, #2 and #3. Sandbox build will fail in integration test (unit tests are skipped).
3. Push all commits. Integration tests should pass (unit tests are skipped).
4. Remove drop commits (unit test + integration tests pass)

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I've added my comments in a drop commit. You do not need to retain that commit.

On top of those comments, a change in plan: since there are a few important unit tests that still depend on AZUL_DSS_ENDPOINT, we should not skip them but instead define that variable in setUp/setUpClass of a suitable test base class.

**Comment by hannes-ucsc** [comment]
Also note that tests didn't fail as expected after pushing commit 1.

File: test/service/test_portal_service.py [hunk 1]
```diff
@@ -41,6 +41,10 @@ def setUpModule():
     configure_test_logging()
 
 
+# TODO: Do not want to lose this one
```
Comment (amarjandu) [inline test/service/test_portal_service.py:44]: Collateral part of #2611

File: test/service/__init__.py [hunk 2]
```diff
@@ -92,3 +97,20 @@ def _fill_index(cls, num_docs=1000):
                                                       aggregate=True),
                            doc_type='meta',
                            refresh='wait_for')
+
+
+class DSSTestCase(TestCase):
+    """
+    This TestCase should be used as a mixin to provide DSS related environment
+    variables to testcases that would break if the variables were not set.
```
Comment (hannes-ucsc) [inline test/service/__init__.py:105]: Slightly awkward.

```suggestion
    A mixin for test cases that depend on certain DSS-related environment
    variables.
```

File: test/service/__init__.py [hunk 3]
```diff
@@ -92,3 +97,20 @@ def _fill_index(cls, num_docs=1000):
...
+
+    @classmethod
+    def setUpClass(cls):
+        cls._catalog_mock = mock.patch.dict(os.environ,
+                                            AZUL_DSS_ENDPOINT='https://dss.data.humancellatlas.org/v1')
+        cls._catalog_mock.start()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls._catalog_mock.stop()
```
Comment (hannes-ucsc) [inline test/service/__init__.py:116]: image

File: test/service/__init__.py [hunk 4]
```diff
@@ -92,3 +97,20 @@ def _fill_index(cls, num_docs=1000):
                                                       aggregate=True),
                            doc_type='meta',
                            refresh='wait_for')
+
+
+class DSSTestCase(TestCase):
```
Comment (hannes-ucsc) [inline test/service/__init__.py:102]:
```suggestion
class DSSUnitTestCase(TestCase):
```

File: test/integration_test.py [hunk 5]
```diff
@@ -681,6 +681,8 @@ def test_openapi(self):
         validate_spec(spec)
 
 
+@unittest.skipIf(config.dss_endpoint is None,
+                 'DSS endpoint is not specified, skipping tests')
```
Comment (hannes-ucsc) [inline test/integration_test.py:685]:
```suggestion
                 'DSS endpoint is not configured')
```

Comment (hannes-ucsc) [reply test/integration_test.py:685]: and elsewhere

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Please chain your PR on #2636 and remove the cherry-picked commit.

File: src/azul/__init__.py [hunk 6]
```diff
@@ -167,8 +167,8 @@ def data_portal_name(self):
         return f'{self.resource_prefix}-data-portal-{self.deployment_stage}'
 
     @property
-    def dss_endpoint(self) -> str:
-        return os.environ['AZUL_DSS_ENDPOINT']
+    def dss_endpoint(self) -> Optional[str]:
+        return os.environ.get('AZUL_DSS_ENDPOINT', None)
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply src/azul/__init__.py:171]: It seems like there is precedence in the codebase for explicitly defining a default, even the default is `None`. 
https://github.com/DataBiosphere/azul/blob/64900c1088e834087c69e5cad4f7f830e86403f5/src/azul/service/hca_response_v5.py#L367

Is that precedent for this section of the codebase? or just something that has not been updated?

File: src/azul/indexer/lambda_iam_policy.py [hunk 7]
```diff
@@ -80,15 +80,19 @@
...
+                {
+                    "Effect": "Allow",
+                    "Action": [
+                        "s3:GetObject",
+                    ],
+                    "Resource": [
+                        f"arn:aws:s3:::{aws.dss_main_bucket(config.dss_endpoint)}/*",
+                    ]
+                },
+            ] if config.dss_endpoint is not None else []
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply src/azul/indexer/lambda_iam_policy.py:94]: was trying to follow similar precedence. https://github.com/DataBiosphere/azul/blob/4b74afda204fbf018f70b8d1932a539c902f1673/src/azul/indexer/lambda_iam_policy.py#L122

File: test/service/__init__.py (hunk 3 above)
Comment (amarjandu) [reply test/service/__init__.py:116]: the name needs to be changed too. >.>

**Comment by hannes-ucsc** [comment]
No need to chain anymore.

File: src/azul/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply src/azul/__init__.py]: Neither. It's clearly a mistake. We should NOT repeat a mistake made in completely different module for the sake of precedent.

File: src/azul/indexer/lambda_iam_policy.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply src/azul/indexer/lambda_iam_policy.py]: That makes sense. Whatever you have is ok.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
Pushed first commit (c1035c1c) to Gitlab. Unit tests should fail.

**Comment by hannes-ucsc** [comment]
It does:

image

https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/8672

**Comment by hannes-ucsc** [comment]
Pushing 1st, 2nd and 3rd commit (c56624cc). IT should fail.

**Comment by hannes-ucsc** [comment]
And it does:

image

https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/8686

**Comment by hannes-ucsc** [comment]
Pushed all fourth (4958ddce) and all preceding commits. IT should pass. UT should be skipped.

**Comment by hannes-ucsc** [comment]
Passes:

image

https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/8696

**Comment by hannes-ucsc** [comment]
The drop! commits demonstrate this sufficiently.

**Pull request merged**